src/base.c

```diff
@@ -491,6 +539,63 @@ base_alloc_edata(tsdn_t *tsdn, base_t *base) {
 	return edata;
 }
 
+static inline void
+b0_alloc_header_size(size_t *header_size, size_t *alignment) {
+	*alignment = QUANTUM;
+	*header_size = QUANTUM > sizeof(edata_t *) ? QUANTUM :
+	    sizeof(edata_t *);
+}
+
+void *
+b0_alloc_tcache_stack(tsdn_t *tsdn, size_t stack_size) {
+	base_t *base = b0get();
+	edata_t *edata = base_alloc_base_edata(tsdn, base);
+	if (edata == NULL) {
+		return NULL;
+	}
+
+	/*
+	 * Reserve room for the header so that edata can be retrieved on dalloc.
+	 * Bump up to usize to improve reusability -- otherwise the freed stacks
+	 * will be put back into the previous size class.
+	 */
+	size_t esn, alignment, header_size;
+	b0_alloc_header_size(&header_size, &alignment);
+
+	size_t alloc_size = sz_s2u(stack_size + header_size);
+	void *addr = base_alloc_impl(tsdn, base, alloc_size, alignment, &esn);
```

Comment:
While `base` is mainly used internally now, the waiting time on `base->mtx` should be rather low? If every thread is using `base->mtx` now during its initialization, will this become a concern?

Reply:
Yeah the base is not accessed frequently, because of 2 factors:
1) metadata is only ever increasing (before this PR), which means once the process reaches peak usage, it won't reach this path anymore
2) the most frequent metadata (re)use, is the `edata_t` -- the [edata_cache](https://github.com/jemalloc/jemalloc/blob/dev/include/jemalloc/internal/edata_cache.h#L16) makes sure they are cached at higher level and only accessing the base when reaching a new peak.

As for the new use case, thread creation is quite heavy weight (and cannot be super frequent) so it won't be a significant factor here.